SpherePacking/ForMathlib/Cusps.lean

```diff
@@ -0,0 +1,50 @@
+import Mathlib.NumberTheory.ModularForms.Cusps
+import Mathlib.NumberTheory.ModularForms.CongruenceSubgroups
+
+open scoped CongruenceSubgroup MatrixGroups
+
+theorem smul_infty_eq_cusp_gamma_one {c : OnePoint ℝ}
+    (hc : IsCusp c (Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(1))) :
+     ∃ A : Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(1),
+       A • OnePoint.infty = c := by
+  rw [Subgroup.IsArithmetic.isCusp_iff_isCusp_SL2Z
+      (Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(1))] at hc
+  rw [isCusp_SL2Z_iff'] at hc
+  obtain ⟨A, hA⟩ := hc
+  rw [Subtype.exists]
+  use A
+  have h1 : Matrix.SpecialLinearGroup.toGL
+     ((Matrix.SpecialLinearGroup.map (Int.castRingHom ℝ)) A)
+       ∈ Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(1) := by
+    simp only [Subgroup.mem_map]
+    exact ⟨A, CongruenceSubgroup.mem_Gamma_one A, rfl⟩
+  use h1
+  symm at hA
+  have : Matrix.SpecialLinearGroup.toGL
+    ((Matrix.SpecialLinearGroup.map (Int.castRingHom ℝ)) A) =
+      (Matrix.SpecialLinearGroup.mapGL ℝ) A := rfl
+  simp [this, hA]
+
+-- TODO: if this theorem is actually true with no additional hypotheses, then
+-- we don't need the above theorem, as it would be a special case.
+theorem smul_infty_eq_cusp_gamma {c : OnePoint ℝ} {N : ℕ}
+    (hc : IsCusp c (Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(N))) :
+     ∃ A : Subgroup.map (Matrix.SpecialLinearGroup.mapGL ℝ) Γ(N),
+       A • OnePoint.infty = c := by
```

Comment:
This theorem is the last missing piece for this PR.

Maybe someone who is more familiar with the math can comment on whether they expect this theorem to be true as stated, and what a proof might look like.

Reply:
We need something like this theorem to deal with https://github.com/leanprover-community/mathlib4/pull/26651